File: src/main/java/phoenix/client/gui/diaryPages/DiaryBook.java
package phoenix.client.gui.diaryPages;

import net.minecraft.client.gui.FontRenderer;
import net.minecraft.client.gui.screen.inventory.ContainerScreen;
import phoenix.client.gui.diaryPages.elements.ADiaryChapter;
import phoenix.client.gui.diaryPages.elements.ADiaryElement;
import phoenix.containers.DiaryContainer;

import java.util.ArrayList;

public class DiaryBook
{
    public final ADiaryChapter current_chapter;
    final int xSize, ySize;
    final FontRenderer font;

    public DiaryBook(int xSizeIn, int ySizeIn, FontRenderer renderer)
    {
        this.xSize = xSizeIn;
        this.ySize = (int) (ySizeIn * 0.6);
        this.font = renderer;
        current_chapter = new ADiaryChapter(xSizeIn, (int) (ySizeIn * 0.6));
    }

    public void next()
    {
        if(!current_chapter.isLast())
            current_chapter.next();
    }

    public void prev()
    {
        if(!current_chapter.isFirst())
            current_chapter.prev();
    }

    public void add(ArrayList<ADiaryElement> elements)
    {
        current_chapter.add(elements);
    }

    public void render(ContainerScreen<DiaryContainer> gui, FontRenderer renderer, int xSize, int ySize, int x, int y, int depth)
    {
        ArrayList<ADiaryElement> page = current_chapter.getCurrentPage1();;
        int sum = 0;
        for (ADiaryElement element : page)
        {
            element.render(gui, renderer, xSize, ySize, x, y + sum * (font.FONT_HEIGHT + 2), depth);
            sum += element.getHeight();
        }


        page = current_chapter.getCurrentPage2();;
        sum = 0;
        for (ADiaryElement element : page)
        {
            element.render(gui, renderer, xSize, ySize, x + xSize / 2 - 10, y + sum * (font.FONT_HEIGHT + 2), depth);
            sum += element.getHeight();
        }
    }
    /*
    public CompoundNBT serialize()
    {
        CompoundNBT res = new CompoundNBT();
        int number = 0;
        for (ArrayList<IDiaryElement> page : pages)
        {
            for (IDiaryElement iDiaryElement : page)
            {
                res.put(number + "", iDiaryElement.serialize());
                number++;
            }
        }
        res.putInt("size", number);
        return res;
    }
    //*/
}

File: src/main/java/phoenix/client/gui/diaryPages/DiaryBook.kt
package phoenix.client.gui.diaryPages

import net.minecraft.client.gui.FontRenderer
import net.minecraft.client.gui.screen.inventory.ContainerScreen
import phoenix.client.gui.diaryPages.elements.ADiaryElement
import phoenix.client.gui.diaryPages.elements.DiaryPage
import phoenix.containers.DiaryContainer
import java.util.*

class DiaryBook(private val xSize: Int, private val ySize: Int, private val font: FontRenderer)
{
    private val pages = LinkedList<DiaryPage>()
    private var page = 0

    fun add(elements: kotlin.collections.ArrayList<ADiaryElement>)
    {
        var i = 0
        while (i in 0 until elements.size)
        {
            if(pages.size == 0)
                pages.add(DiaryPage())
            while (i < elements.size && pages.last.tryAdd(elements[i], xSize, ySize))
            {
                i++
            }
            pages.add(DiaryPage())
        }
    }

    fun add(element: ADiaryElement) = add(arrayListOf(element))

    fun render(gui: ContainerScreen<DiaryContainer>, renderer: FontRenderer, xSize: Int, ySize: Int, x: Int, y: Int, depth: Int)
    {
        leftPage.render(gui, font, xSize / 2, ySize, x, y, depth)
        rightPage.render(gui, renderer, xSize / 2, ySize, x + xSize / 2, y, depth)
    }

    private val isLast : Boolean get() = page + 1 >= pages.size - 1
    private val isFirst: Boolean get() = page     <= 0
    private val leftPage: DiaryPage  get() = if(page     in 0 until pages.size) pages[page]     else DiaryPage()
    private val rightPage: DiaryPage get() = if(page + 1 in 0 until pages.size) pages[page + 1] else DiaryPage()

    fun next() = if (!isLast)  page += 2 else page += 0
    fun prev() = if (!isFirst) page -= 2 else page += 0
}